libraries/ms-common/src/x86/interrupt.c

```diff
@@ -0,0 +1,151 @@
+  sigaddset(&mask, SIGRTMIN + INTERRUPT_PRIORITY_LOW);
+  sigaddset(&mask, SIGRTMIN + INTERRUPT_PRIORITY_NORMAL);
+  sigaddset(&mask, SIGRTMIN + INTERRUPT_PRIORITY_HIGH);
+  sigprocmask(SIG_SETMASK, &mask, NULL);
+}
+
+// REGISTRATION AND TRIGGER HELPERS
+// Use these functions to register and trigger interrupts using module specific registration and
+// trigger functions.
+
+// Register an interrupt and assign it an id. When the interrupt is triggered via a software
+// interrupt of external signal it must be referenced by this id the sival_int.
+StatusCode prv_register_interrupt(InterruptSettings *settings, interrupt_callback callback,
+                                  uint8_t *id) {
+  if (s_interrupt_next_id >= NUM_INTERRUPTS) {
+    return status_msg(STATUS_CODE_RESOURCE_EXHAUSTED, "Too many interrupts registered.");
+  } else if (settings->type >= NUM_INTERRUPT_TYPE || settings->edge >= NUM_INTERRUPT_EDGE ||
+             settings->priority >= NUM_INTERRUPT_PRIORITY) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // Assign an interrupt to an array and give it a new id.
+  s_interrupt_map[s_interrupt_next_id].priority = settings->priority;
+  s_interrupt_map[s_interrupt_next_id].callback = callback;
+  *id = s_interrupt_next_id++;
+
+  return STATUS_CODE_OK;
+}
+
+// Trigger a software interrupt for the given interrupt by its id.
+StatusCode prv_trigger_interrupt(uint8_t id) {
+  if (id == INTERRUPT_UNUSED) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // Enqueue a new signal sent to this process that has a signal number determined by the id for the
+  // callback it is going to run.
+  siginfo_t value_store;
+  value_store.si_value.sival_int = id;
+  sigqueue(getpid(), SIGRTMIN + s_interrupt_map[id].priority, value_store.si_value);
+  return STATUS_CODE_OK;
+}
+
+// GPIO INTERRUPTS
+
+#define NUM_GPIO_INTERRUPTS 16
+static uint8_t s_interrupt_gpio_id_map[NUM_GPIO_INTERRUPTS];
+
+StatusCode interrupt_gpio_register(GPIOAddress *address, InterruptSettings *settings,
+                                   interrupt_callback callback) {
+  if (address->pin >= NUM_GPIO_PORTS || address->port >= NUM_GPIO_PINS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  } else if (s_interrupt_gpio_id_map[address->pin] != INTERRUPT_UNUSED) {
+    return status_msg(STATUS_CODE_RESOURCE_EXHAUSTED, "Interrupt for this pin is already used.");
+  }
+
+  status_ok_or_return(
+      prv_register_interrupt(settings, callback, &s_interrupt_gpio_id_map[address->pin]));
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode interrupt_gpio_trigger(GPIOAddress *address) {
+  if (address->pin >= NUM_GPIO_PORTS || address->port >= NUM_GPIO_PINS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  } else if (s_interrupt_gpio_id_map[address->pin] == INTERRUPT_UNUSED) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "No interrupt set for this pin.");
+  }
+
+  status_ok_or_return(prv_trigger_interrupt(s_interrupt_gpio_id_map[address->pin]));
```

Comment:
Couldn't you just return the value you get from prv_trigger_interrupt?
I think nested function calls should be avoided for the sake of readability :P